File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

	class Node {
		int number;
		int distance;

		public Node(int number, int distance) {
			super();
			this.number = number;
			this.distance = distance;
		}

	}

	int V, E, K;
	ArrayList<ArrayList<Node>> graph = new ArrayList<>();
	int[] distances;
	final int INF = Integer.MAX_VALUE;

	public void solution() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();

		V = Integer.valueOf(st.nextToken());
		E = Integer.valueOf(st.nextToken());

		K = Integer.valueOf(br.readLine());

		for (int i = 0; i < V + 1; i++) {
			graph.add(new ArrayList<>());
		}

		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int from = Integer.valueOf(st.nextToken());
			int to = Integer.valueOf(st.nextToken());
			int distance = Integer.valueOf(st.nextToken());

			graph.get(from)
				.add(new Node(to, distance));
		}

		distances = new int[V + 1];
		Arrays.fill(distances, INF);

		dijkstra(K);

		for (int i = 1; i < distances.length; i++) {
			if (distances[i] == INF) {
				sb.append("INF")
					.append("\n");

			} else {
				sb.append(distances[i])
					.append("\n");
			}
		}
		
		System.out.println(sb);
	}

	private void dijkstra(int start) {
		PriorityQueue<Node> pq = new PriorityQueue<>((o1, o2) -> o1.distance - o2.distance);
		pq.offer(new Node(start, 0));
		distances[start] = 0;

		while (!pq.isEmpty()) {
			// 최소 우선순위큐 에서 하나 꺼냄
			Node minVertex = pq.poll();
			int current = minVertex.number;
			int distance = minVertex.distance;

			// 큐에서 꺼낸 거리와, 최단 거리 테이블 값을 비교해서 방문처리를 함
			// 만약 현재 꺼낸 노드의 거리가 최단 거리 테이블보다 값이 크다면 해당 노드는 이전에 방문된 노드
			if (distance > distances[current]) {
				continue;
			}

			// 큐에서 꺼낸 노드에서 연결된 노드들을 탐색
			for (Node linkedNode : graph.get(current)) {
				// 해당 노드를 거쳐서 다음 노드로 이동 할 때의 값이 다음 이동 노드의 최단거리 테이블보다 작을 때
				if (distance + linkedNode.distance < distances[linkedNode.number]) {
					// 값 갱신
					distances[linkedNode.number] = distance + linkedNode.distance;
					// 갱신된 노드를 pq에 넣음
					pq.offer(new Node(linkedNode.number, distances[linkedNode.number]));
				}
			}
		}

	}

	public static void main(String[] args) throws IOException {
		new Main().solution();
	}
}

File: 백준/Gold/1753. 최단경로/최단경로.js
const input = require('fs').readFileSync(process.version === 'v20.9.0' ? 'input.txt' : '/dev/stdin')
  .toString().trim().split("\n");


class PriorityQueue {
  constructor() {
    this.values = [];
  }

  swap(a, b) {
    [this.values[a], this.values[b]] = [this.values[b], this.values[a]];
  }

  bubbleUp() {
    let index = this.values.length - 1;
    let el = this.values[index];

    while (index > 0) {
      let pi = Math.floor((index - 1) / 2);
      let p = this.values[pi];

      if (el.w >= p.w) break;

      this.swap(index, pi);
      index = pi;
    }
  }

  eq(v, w) {
    const newNode = {v, w};
    this.values.push(newNode);
    this.bubbleUp();
  }

  sinkDown() {
    let index = 0;
    let length = this.values.length;
    let el = this.values[index];

    while (true) {
      let lI = index * 2 + 1;
      let rI = index * 2 + 2;
      let left, right;
      let swapIndex = null;

      if (lI < length) {
        left = this.values[lI];
        if (left.w < el.w) {
          swapIndex = lI;
        }
      }

      if (rI < length) {
        right = this.values[rI];
        if ((swapIndex === null && right.w < el.w) ||
          (swapIndex !== null && right.w < left.w)) {
          swapIndex = rI;
        }
      }

      if (swapIndex === null) break;

      this.swap(index, swapIndex);
      index = swapIndex;
    }
  }

  dq() {
    let head = this.values[0];
    let popped = this.values.pop();

    if (this.values.length > 0) {
      this.values[0] = popped;
      this.sinkDown();
    }

    return head;
  }

  size() {
    return this.values.length;
  }

}


const solution = () => {
  const [N, M] = input[0].split(' ').map(Number);
  const K = +input[1];
  const graph = {};

  for (let i = 1; i <= N; i++) {
    graph[i] = [];
  }

  for (let i = 2; i < M + 2; i++) {
    const [start, to, weight] = input[i].split(" ").map(Number);
    graph[start].push({to, weight});
  }

  const distances = dijkstra(K)
  const result = [];

  for(let i = 1; i <= N; i++){
    if(distances[i] === Infinity){
      result.push("INF");
    } else {
      result.push(distances[i]);
    }
  }

  console.log(result.join("\n"));


  function dijkstra(start) {
    const pq = new PriorityQueue();
    const distances = Array(N + 1).fill(Infinity);

    pq.eq(start, 0);
    distances[start] = 0;

    while (pq.size() > 0) {
      // 1. 우선순위 큐에서 가장 작은 가중치를 가진 정점을 꺼냅니다.
      const {v: current, w} = pq.dq();

      // 2. 현재 정점까지의 최단 거리가 이미 꺼낸 가중치보다 작다면,
      // 더 짧은 경로가 존재한다는 뜻이므로 이 정점은 무시합니다.
      if (distances[current] < w) continue;

      // 3. 현재 정점의 모든 이웃을 순회합니다.
      graph[current].forEach(neighbor => {
        // 4. 현재 정점을 거쳐 이웃으로 가는 총 가중치를 계산합니다.
        let totalWeight = w + neighbor.weight;

        // 5. 현재 정점을 거쳐 이웃으로 가는 거리가 현재 알려진 이웃까지의 거리보다 작다면,
        //    이 거리를 업데이트하고 우선순위 큐에 넣습니다.
        if (totalWeight < distances[neighbor.to]) {
          distances[neighbor.to] = totalWeight;
          pq.eq(neighbor.to, totalWeight);
        }
      });
    }

    return distances;
  }

};

solution();